optaplanner-examples/src/main/java/org/optaplanner/examples/common/experimental/impl/ConsecutiveSetTree.java

```diff
@@ -0,0 +1,329 @@
+            BiFunction<DifferenceType_, DifferenceType_, DifferenceType_> sumFunction,
+            DifferenceType_ maxDifference,
+            DifferenceType_ zeroDifference) {
+        this.indexFunction = indexFunction;
+        this.differenceFunction = differenceFunction;
+        this.sumFunction = sumFunction;
+        this.maxDifference = maxDifference;
+        this.zeroDifference = zeroDifference;
+        // Identity Hashcode for duplicate protection
+        // Ex: two different games on the same time slot
+        comparator = Comparator.comparing(indexFunction).thenComparing(System::identityHashCode);
+        itemToCountMap = new IdentityHashMap<>();
+        itemSet = new TreeSet<>(comparator);
+        startItemToSequence = new TreeMap<>(comparator);
+        startItemToPreviousBreak = new TreeMap<>(comparator);
+        consecutiveData = new ConsecutiveDataImpl<>(this);
+        sequenceList = new MapValuesIterable<>(startItemToSequence);
+        breakList = new MapValuesIterable<>(startItemToPreviousBreak);
+    }
+
+    protected NavigableSet<ValueType_> getItemSet() {
+        return itemSet;
+    }
+
+    public Comparator<ValueType_> getComparator() {
+        return comparator;
+    }
+
+    public Iterable<SequenceImpl<ValueType_, DifferenceType_>> getConsecutiveSequences() {
+        return sequenceList;
+    }
+
+    public Iterable<BreakImpl<ValueType_, DifferenceType_>> getBreaks() {
+        return breakList;
+    }
+
+    public void updateLengthOfBreak(BreakImpl<ValueType_, DifferenceType_> theBreak) {
+        theBreak.setLength(getBreakLengthBetween(theBreak.getAfterItem(), theBreak.getBeforeItem()));
+    }
+
+    public DifferenceType_ getSequenceLength(Sequence<ValueType_, DifferenceType_> sequence) {
+        return sumFunction.apply(maxDifference, differenceFunction.apply(indexFunction.apply(sequence.getFirstItem()),
+                indexFunction.apply(sequence.getLastItem())));
+    }
+
+    public DifferenceType_ getBreakLengthBetween(ValueType_ from, ValueType_ to) {
+        return differenceFunction.apply(indexFunction.apply(from), indexFunction.apply(to));
+    }
+
+    public ConsecutiveDataImpl<ValueType_, DifferenceType_> getConsecutiveData() {
+        return consecutiveData;
+    }
+
+    public ValueType_ getEndItem(ValueType_ key) {
+        return startItemToSequence.get(key).getLastItem();
+    }
+
+    private boolean isFirstSuccessorOfSecond(PointType_ first, ValueType_ firstValue, PointType_ second,
+            ValueType_ secondValue) {
+        DifferenceType_ difference = differenceFunction.apply(second, first);
+        int differenceComparedTo0 = difference.compareTo(zeroDifference);
+        if (differenceComparedTo0 > 0) {
+            return difference.compareTo(maxDifference) <= 0;
+        } else {
+            return differenceComparedTo0 == 0 && System.identityHashCode(secondValue) < System.identityHashCode(firstValue);
+        }
```

Comment:
Important: all comparisons *must* happen in a shared piece of code. Especially because of this identity trick.
This code all be put in one place, and be well documented there - so that if/when we're touching this in the future, we understand the trick and can rely on the fact that this trick is not exposed anywhere else.

Reply:
Can't be put into a Comparator; it uses difference types, and the items aren't available within the type; did it within a generic function that takes four parameters